### DISCUSSION TITLE
how BleskOS identify a newly plugged-in device?

### DISCUSSION THREAD

**Discussion by davidgaosheng** [VendelinSlezak/BleskOS#20] [Q&A] [answered]
hello!
A usb keyboard were plugged in. 
PIT timer interrupt handler calls _run_processes_on_background_, which  calls _detect_status_change_of_usb_devices_ when ticks_of_processes==0x80, which calls
_detect_status_change_of_usb_devices_, which calls _uhci_controller_detect_status_change_, where I hope the newly plugged in usb keyboard would be taken care of. However 
_uhci_controller_detect_status_change_ doesn't call _uhci_controller_detect_devices_ as I wished. Had I missed something? Thanks

**Comment by VendelinSlezak** [comment, marked as answer]
*detect_status_change_of_usb_devices()* only search for change of connect status. If it found one, it set variabile usb_new_device_detected to 1. It is because configuring new USB device must have some timeout in case of issues, and waiting is held by PIT interrupt, which is not fired when we execute its irq handler. So PIT interrupt just found if there was some connection change, and it is matter of other code to detect that change took place and run code to identify device that was plugged/unplugged. Now this take place in method *wait_for_usb_mouse()*, where we wait not only for usb mouse, but also for any input from user - plugging usb device is considered as input also - so when usb_new_device_detected changes to 1, routine *wait_for_usb_mouse()* calls *detect_usb_devices()* and returns, so program that called this routine can also show new connection. You can found this code on https://github.com/VendelinSlezak/BleskOS/blob/master/source/drivers/usb/usb_mouse.c#L43
